src/document/generator.ml

```diff
@@ -1783,26 +1799,152 @@ module Make (Syntax : SYNTAX) = struct
+              | (name2, Dir d) :: q when name = name2 ->
+                  (name2, Dir (add_entry_to_dir d payload rest)) :: q
+              | d :: q -> d :: add_to_dir q
+            in
+            (p, add_to_dir l)
+
+      let make l =
+        let empty = (None, []) in
+        let add_entry_to_dir acc (path, payload) =
+          add_entry_to_dir acc path payload
+        in
+        List.fold_left add_entry_to_dir empty l
+
+      let rec remove_common_root = function
+        | None, [ (_, Dir d) ] -> remove_common_root d
+        | x -> x
+
+      let rec to_sidebar ?(fallback = "root") convert (name, content) =
+        let name =
+          match name with
+          | Some v -> convert v
+          | None -> block (Block.Inline [ inline (Text fallback) ])
+        in
+        let content =
+          let content = List.map (t_to_sidebar convert) content in
+          block (Block.List (Block.Unordered, content))
+        in
+        [ name; content ]
+
+      and t_to_sidebar convert = function
+        | _, Leaf payload -> [ convert payload ]
+        | fallback, Dir d -> to_sidebar ~fallback convert d
+    end
+
+    let sidebar root_id (v : Odoc_model.Lang.Sidebar.t) =
+      let root_id = (root_id :> Paths.Identifier.t) in
+      let prepare (link_content, x) =
+        let x = (x :> Paths.Identifier.t) in
+        let payload =
+          (link_content, x, Odoc_model.Paths.Identifier.equal x root_id)
+        in
+        let path = Odoc_model.Paths.Identifier.fullname x in
+        (payload, path)
+      in
+      let title t =
+        block
+          (Inline [ inline (Inline.Styled (`Bold, [ inline (Inline.Text t) ])) ])
+      in
+      let page_hierarchy { Odoc_model.Lang.Sidebar.page_name; pages } =
+        if pages = [] then []
+        else
+          let pages = List.map prepare pages in
+          let hierarchy =
+            Hierarchy.make pages |> Hierarchy.remove_common_root
+          in
+          let convert (content, id, is_highlighted) =
+            let url = Url.from_identifier ~stop_before:false id in
+            match url with
+            | Ok href ->
+                let target = InternalLink.Resolved href in
+                let content = Comment.link_content content in
+                let link = { InternalLink.target; content; tooltip = None } in
+                let attr = if is_highlighted then [ "current_unit" ] else [] in
+                block (Inline [ inline ~attr @@ Inline.InternalLink link ])
+            | Error _ ->
+                let content = Comment.link_content content in
+                (* let attr = if is_highlighted then [ "current_unit" ] else [] in *)
+                block (Inline content)
+          in
+
+          let pages = Hierarchy.to_sidebar convert hierarchy in
+          let pages = [ block (Block.List (Block.Unordered, [ pages ])) ] in
+          [ title @@ page_name ^ "'s Pages" ] @ pages
+      in
+      let rec concat_map acc f = function
```

Comment:
You've defined `concat_map` for the 6th time and `filter_map` for the 5th time in this PR.
We have 4 `Utils` modules in the projects, I think it's time for a unified "utils" library.

Reply:
I can do that but it will introduce unrelated diffs in this PR. I'll try and we'll see.

Reply:
It can be done in a later PR!